tests/test_library.py

```diff
@@ -1,303 +1,360 @@
-
-    tidal_search = mocker.Mock()
-    tidal_search.return_value = ([artist], [], [])
-    session.artist.return_value = artist
-    mocker.patch("mopidy_tidal.search.tidal_search", tidal_search)
-    res = tlp.get_distinct(field, query={"any": "any"})
-    tidal_search.assert_called_once_with(session, query={"any": "any"}, exact=True)
-    session.artist.assert_called_once_with("1")
-    artist.get_albums.assert_called_once_with()
-    assert len(res) == 1
-    assert res[0] == "Thing [TIDAL]"
-
-
-@pytest.mark.parametrize("field", ("album", "albumartist"))
-def test_get_distinct_album_new_api_no_artist(tlp, mocker, field):
-    tlp, backend = tlp
-    session = backend.session
-    session.mock_add_spec(("tidal_search", "artist", "artist.get_albums"))
-
-    artist = mocker.Mock()
-    artist.uri = "tidal:artist:1"
-
-    tidal_search = mocker.Mock()
-    tidal_search.return_value = ([artist], [], [])
-    session.artist.return_value = None
-    mocker.patch("mopidy_tidal.search.tidal_search", tidal_search)
-    assert not tlp.get_distinct(field, query={"any": "any"})
-    tidal_search.assert_called_once_with(session, query={"any": "any"}, exact=True)
-    session.artist.assert_called_once_with("1")
-
-
-def test_browse_wrong_uri(tlp):
-    tlp, backend = tlp
-    assert not tlp.browse("")
-    assert not tlp.browse("spotify:something:something_else")
-    assert not tlp.browse("tidal:album:oneid:oneidtoomany")
-
-
-def test_browse_root(tlp):
-    tlp, backend = tlp
-    assert tlp.browse("tidal:directory") == [
-        Ref(name="Genres", type="directory", uri="tidal:genres"),
-        Ref(name="Moods", type="directory", uri="tidal:moods"),
-        Ref(name="Mixes", type="directory", uri="tidal:mixes"),
-        Ref(name="My Artists", type="directory", uri="tidal:my_artists"),
-        Ref(name="My Albums", type="directory", uri="tidal:my_albums"),
-        Ref(name="My Playlists", type="directory", uri="tidal:my_playlists"),
-        Ref(name="My Tracks", type="directory", uri="tidal:my_tracks"),
-    ]
-
-
-def test_browse_artists(tlp, mocker, tidal_artists):
-    tlp, backend = tlp
-    session = backend.session
-    session.user.favorites.artists = tidal_artists
-    mocker.patch("mopidy_tidal.library.get_items", lambda x: x)
-    assert tlp.browse("tidal:my_artists") == [
-        Ref(name="Artist-0", type="artist", uri="tidal:artist:0"),
-        Ref(name="Artist-1", type="artist", uri="tidal:artist:1"),
-    ]
-
-
-def test_browse_albums(tlp, mocker, tidal_albums):
-    tlp, backend = tlp
-    session = backend.session
-    session.user.favorites.albums = tidal_albums
-    mocker.patch("mopidy_tidal.library.get_items", lambda x: x)
-    assert tlp.browse("tidal:my_albums") == [
-        Ref(name="Album-0", type="album", uri="tidal:album:0"),
-        Ref(name="Album-1", type="album", uri="tidal:album:1"),
-    ]
 
+class TestSearch:
+    def test_defers_to_tidal_search(self, library_provider, mocker):
```

Comment:
Defer in this context = ?

Reply:
I use `defer` when I test that some functionality is just handed over to another unit we test somewhere else.  So if `foo(x)` just calls `bar(x)` (and the fact that `foo` calls `bar` is part of the contract) the only behaviour we can assert about `foo` is that it hands the right data over to `bar` and then returns the result.

Concretely here, search is simply a wrapper around `python_tidal`'s search.  The only thing it should do is format the query, hand it to `python_tidal`, format the result and then hand it out. The actual searching is deferred to the underlying library, and we shouldn't try to test it.

It's just a word I use; don't think you'll find any blog posts on it. There's probably a proper term for it.

Contrast this to testing that e.g. `TestBrowse:test_invalid_uri_returns_empty_list`.  That behaviour is part of the spec of `browse()`, so we just hand in an invalid uri and check that we get an empty list back---the particular details of how that happens aren't part of the spec and shouldn't be tested.  If `browse` call some internal function, that's its business.

Reply:
Thanks for the detailed explanation, that makes good sense.